workflow/controller/operator.go

```diff
@@ -1037,6 +1037,36 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatu
 			node.Phase = wfv1.NodeError
 		} else {
 			node.Outputs = &outputs
+			tmplCtx, err := woc.createTemplateContext(node.GetTemplateScope())
+			if err != nil {
+				log.Errorf("Failed to create template context for node %s", node.ID)
+				return nil
+			}
+			_, resolvedTmpl, _, err := tmplCtx.ResolveTemplate(node)
+			if err != nil {
+				log.Errorf("Failed to resolve template for node %s: %s", node.ID, err)
+				return nil
+			}
+
+			localParams := make(map[string]string)
+			_, localArgs, _ := woc.loadExecutionSpec()
+			woc.setGlobalParameters(localArgs)
+			// Inject the pod name. If the pod has a retry strategy, the pod name will be changed and will be injected when it
+			// is determined
+			if resolvedTmpl.IsPodType() && resolvedTmpl.RetryStrategy == nil {
+				localParams[common.LocalVarPodName] = woc.wf.NodeID(node.Name)
+			}
+
+			tmpl, err := common.ProcessArgs(resolvedTmpl, &localArgs, woc.globalParams, localParams, false)
+			if err != nil {
+				log.Errorf("Failed to process template for node %s: %s", node.ID, err)
+			}
+
+			if tmpl.Memoize != nil {
+				c := woc.controller.cache.(*configMapCache)
```

Comment:
Why are we casting here? The whole idea of using the interface is that we don't need to know this is a `configMapCache`.

Reply:
The issue here is that we have to set the `configMapName` here, and since not every cache type has a `configMapName` there cannot be a generic cache struct that implements the `MemoizationCache` interface, and so to set this information the cast is necessary.

To make the controller completely unaware of the cache type, the cache needs to be aware of the template. I'll explore passing the template to the cache on initialization.